File: src/jp/senchan/android/wasatter/adapter/Timeline.java
package jp.senchan.android.wasatter.adapter;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

import jp.senchan.android.wasatter.R;
import jp.senchan.android.wasatter.Setting;
import jp.senchan.android.wasatter.Wasatter;
import jp.senchan.android.wasatter.client.WassrUrl;
import jp.senchan.android.wasatter.item.Item;
import jp.senchan.android.wasatter.setting.DisplaySetting;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.text.Html;
import android.text.SpannableStringBuilder;
import android.text.Html.ImageGetter;
import android.text.TextUtils.TruncateAt;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

public class Timeline extends ArrayAdapter<Item> implements
		WasatterAdapter {

	private ArrayList<Item> items;
	private LayoutInflater inflater;

	public Timeline(Context context, int textViewResourceId,
			ArrayList<Item> items, boolean channel) {
		super(context, textViewResourceId, items);
		this.items = items;
		this.inflater = (LayoutInflater) context
				.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
	}

	public View getView(int position, View view, ViewGroup parent) {
		if (view == null) {
			// 受け取ったビューがnullなら新しくビューを生成
			view = this.inflater.inflate(R.layout.timeline_row, null);
		}
		// データの取得
		Item item = this.items.get(position);
		if (item != null) {
			TextView screenName = (TextView) view
					.findViewById(R.id.screen_name);

			// 名前をビューにセットする
			TextView text = (TextView) view.findViewById(R.id.status);
			if (screenName != null) {
				screenName.setText(item.name);
			}
			// テキストをビューにセットする
			if (text != null) {
				CharSequence html = Html.fromHtml(item.html,
						new ImageGetter() {

							
							public Drawable getDrawable(String source) {
								// 必要な画像のURLをあらかじめ取得
								Bitmap bmp = Wasatter.images
										.get(source);
								BitmapDrawable bd = new BitmapDrawable(
										bmp);
								// TODO:解像度ごとにサイズ変えられたらいいなああああ
								Rect bounds = new Rect(0, 0, 20, 20);
								bd.setBounds(bounds);
								return bd;
							}
						}, null);
				text.setText(html);
			}
			// テキストの行数を決定する。
			boolean multiLine = DisplaySetting.get(DisplaySetting.BODY_MULTILINE, true);
			text.setSingleLine(!multiLine);
			if (!multiLine) {
				text.setEllipsize(TruncateAt.MARQUEE);
			} else {
				text.setEllipsize(null);
			}
			// 返信元の名前をビューにセットする
			TextView reply_name = (TextView) view.findViewById(R.id.reply_name);
			if (reply_name != null && !"null".equals(item.replyUserNick)
					&& item.replyUserNick != null) {
				reply_name.setText(new SpannableStringBuilder(">").append(
						item.replyUserNick).toString());
				reply_name.setVisibility(View.VISIBLE);
			} else if (reply_name != null) {
				reply_name.setText("");
				reply_name.setVisibility(View.GONE);
			}
			// アイコンをロードする。
			Bitmap image = Wasatter.images.get(item.profileImageUrl);
			ImageView icon = (ImageView) view.findViewById(R.id.icon);
			if (!Setting.isLoadImage()) {
				icon.setVisibility(View.GONE);
			} else {
				icon.setImageBitmap(image);
				icon.setVisibility(View.VISIBLE);
			}
			// サービス名をビューにセットする
			TextView service = (TextView) view.findViewById(R.id.service_name);
			if (service != null) {
				service.setText(item.service);
			}
			// イイネ！アイコンリストを表示
			LinearLayout layout_favorite_list = (LinearLayout) view
					.findViewById(R.id.layout_favorite);
			LinearLayout layout_favorite_icons = (LinearLayout) view
					.findViewById(R.id.layout_favorite_images);
			if (item.favorite.size() == 0) {
				layout_favorite_list.setVisibility(View.GONE);
				layout_favorite_icons.removeAllViews();
			} else {
				ArrayList<String> favorites = item.favorite;
				int count = favorites.size();
				TextView tv = new TextView(view.getContext());
				tv.setText(new SpannableStringBuilder("x").append(
						String.valueOf(count)).toString());
				layout_favorite_icons.removeAllViews();
				layout_favorite_icons.addView(tv);
				layout_favorite_list.setVisibility(View.VISIBLE);
				//TODO イイネアイコンのロード設定
				//if (Setting.isLoadFavoriteImage()) {
					for (int i = 0; i < count; i++) {
						ImageView add_icon = new ImageView(view.getContext());
						add_icon.setImageBitmap(Wasatter.images
								.get(WassrUrl.FAVORITE_ICON.replace(
										"[user]", favorites.get(i))));
						add_icon.setLayoutParams(new LayoutParams(28, 28));
						add_icon.setPadding(2, 2, 2, 2);
						layout_favorite_icons.addView(add_icon);
					}
				//}
			}

			//投稿日時表示するぜヒャッハー
			TextView date = (TextView) view.findViewById(R.id.post_date);
			SimpleDateFormat sdf = new SimpleDateFormat("yy/MM/dd HH:mm:ss");
			Date now = new Date();
			now.setTime(item.epoch * 1000);
			date.setText(sdf.format(now));
		}
		return view;
	}

	
	public void updateView() {
		// TODO 自動生成されたメソッド・スタブ
		super.notifyDataSetChanged();
	}
}

File: app/src/main/java/jp/senchan/android/wasatter/adapter/Timeline.kt
package jp.senchan.android.wasatter.adapter

import android.content.Context
import android.net.Uri
import android.text.SpannableStringBuilder
import android.text.TextUtils.TruncateAt
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.squareup.picasso.Picasso
import jp.senchan.android.wasatter.R
import jp.senchan.android.wasatter.WasatterAdapter
import jp.senchan.android.wasatter.WasatterItem
import jp.senchan.android.wasatter.repository.SettingsRepository
import java.text.SimpleDateFormat
import java.util.*

class Timeline(context: Context, textViewResourceId: Int,
               private val items: ArrayList<WasatterItem>?, channel: Boolean) : ArrayAdapter<WasatterItem>(context, textViewResourceId, items!!), WasatterAdapter {
    private val inflater: LayoutInflater =  context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    private val settingsRepository by lazy {
        SettingsRepository.getDefaultInstance(context)
    }
    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        val view = view ?: inflater.inflate(R.layout.timeline_row, null)
        // データの取得
        val item = items!![position]
        if (item != null) {
            val screenName = view.findViewById<View>(R.id.screen_name) as TextView
            // 名前をビューにセットする
            val text = view!!.findViewById<View>(R.id.status) as TextView
            if (screenName != null) {
                screenName.text = item.name
            }
            // テキストをビューにセットする
            if (text != null) {
                text.text = item.text
            }
            // テキストの行数を決定する。
            text.isSingleLine = !settingsRepository.isDisplayBodyMultiLine
            if (!settingsRepository.isDisplayBodyMultiLine) {
                text.ellipsize = TruncateAt.MARQUEE
            } else {
                text.ellipsize = null
            }
            // 返信元の名前をビューにセットする
            val reply_name = view.findViewById<View>(R.id.reply_name) as TextView
            if ("null" != item.replyUserNick && item.replyUserNick != null) {
                reply_name.text = SpannableStringBuilder(">").append(
                        item.replyUserNick).toString()
                reply_name.visibility = View.VISIBLE
            } else {
                reply_name.text = ""
                reply_name.visibility = View.GONE
            }
            // 画像をセット
            val icon = view.findViewById<View>(R.id.icon) as ImageView
            if (settingsRepository.isLoadImage) {
                Picasso.get().isLoggingEnabled = true
                Picasso.get()
                        .load(Uri.parse(item.profileImageUrl))
                        .into(icon)
                icon.visibility = View.VISIBLE
            } else {
                icon.visibility = View.GONE
            }
            // サービス名をビューにセットする
            val service = view.findViewById<View>(R.id.service_name) as TextView
            service.text = item.service
            // イイネ！アイコンリストを表示
            val layout_favorite_list = view
                    .findViewById<View>(R.id.layout_favorite) as LinearLayout
            val layout_favorite_icons = view
                    .findViewById<View>(R.id.layout_favorite_images) as LinearLayout
            if (item.favorite!!.size == 0) {
                layout_favorite_list.visibility = View.GONE
                layout_favorite_icons.removeAllViews()
            } else {
                val favorites = item.favorite
                val count = favorites.size
                val tv = TextView(view.context)
                tv.text = SpannableStringBuilder("x").append(count.toString()).toString()
                layout_favorite_icons.removeAllViews()
                layout_favorite_icons.addView(tv)
                layout_favorite_list.visibility = View.VISIBLE
                if (settingsRepository.isLoadFavoriteImage) {
                    for (i in 0 until count) {
                        val add_icon = ImageView(view.context)
                        add_icon.layoutParams = ViewGroup.LayoutParams(28, 28)
                        add_icon.setPadding(2, 2, 2, 2)
                        layout_favorite_icons.addView(add_icon)
                    }
                }
            }
            //投稿日時表示するぜヒャッハー
            val date = view.findViewById<View>(R.id.post_date) as TextView
            val sdf = SimpleDateFormat("yyyy/MM/dd HH:mm:ss")
            val now = Date()
            now.time = item.epoch * 1000
            date.text = sdf.format(now)
        }
        return view!!
    }

    override fun updateView() { // TODO 自動生成されたメソッド・スタブ
        super.notifyDataSetChanged()
    }
}